jepsen/scalardb/src/scalardb/transfer_append.clj

```diff
@@ -0,0 +1,328 @@
+  "Read a record with a transaction. If read fails, this function returns nil."
+  [tx id]
+  (try
+    (first (.scan tx (prepare-scan id)))
+    (catch CrudException _ nil)))
+
+(defn read-all-records
+  "Read records from 0 .. (n - 1)"
+  [test n]
+  (let [tx (scalar/start-transaction test)]
+    (map #(read-record tx %) (range n))))
+
+(defn read-all-with-retry
+  "Read records from 0 .. (n - 1) and retry if needed"
+  [test n]
+  (loop [tries scalar/RETRIES]
+    (when (< tries scalar/RETRIES)
+      (scalar/exponential-backoff (- scalar/RETRIES tries)))
+    (when (zero? (mod tries scalar/RETRIES_FOR_RECONNECTION))
+      (scalar/prepare-transaction-service! test))
+    (let [results (read-all-records test n)]
+      (if (empty? (filter nil? results))
+        results
+        (if (pos? tries)
+          (recur (dec tries))
+          (throw (ex-info "Failed to read all records"
+                          {:cause "Failed to read all records"})))))))
+
+(defn get-balance
+  "Get the balance from a result"
+  [^Result result]
+  (-> result (.getValue BALANCE) .get .get))
+
+(defn get-version
+  "Get the version from a result"
+  [^Result result]
+  (-> result (.getValue scalar/VERSION) .get .get))
+
+(defn get-balances-and-versions
+  "Read all records with a transaction. Return only balances and versions."
+  [test n]
+  (let [results (read-all-with-retry test n)]
+    (map #(assoc {} :balance (get-balance %) :version (get-version %)) results)))
+
+(defn- try-reconnection!
+  [test]
+  (when (= (swap! (:failures test) inc) NUM_FAILURES_FOR_RECONNECTION)
+    (scalar/prepare-transaction-service! test)
+    (reset! (:failures test) 0)))
+
+(defrecord TransferClient [initialized? n initial-balance]
+  client/Client
+  (open! [_ _ _]
+    (TransferClient. initialized? n initial-balance))
+
+  (setup! [_ test]
+    (locking initialized?
+      (when (compare-and-set! initialized? false true)
+        (let [session (drv/connect (->> test :nodes (map name)))]
+          (create-transfer-table! session test)
+          (scalar/create-coordinator-table! session test)
+          (drv/disconnect! session))
+        (scalar/prepare-storage-service! test)
+        (scalar/prepare-transaction-service! test)
+        (populate-accounts test n initial-balance))))
+
+  (invoke! [_ test op]
+    (case (:f op)
+      :transfer (if-let [tx (scalar/start-transaction test)]
+                  (try
+                    (tx-transfer tx (-> op :value :from) (-> op :value :to) (-> op :value :amount))
+                    (assoc op :type :ok)
+                    (catch UnknownTransactionStatusException e
+                      (swap! (:unknown-tx test) conj (.getId tx))
+                      (assoc op :type :fail :error {:unknown-tx-status (.getId tx)}))
+                    (catch UncommittedRecordException e
+                      (assoc op :type :fail :error (:uncommitted-record (.getMessage e))))
+                    (catch NoMutationException e
```

Comment:
Is `NoMutationException` thrown for Scalar DB transaction?
I think it is thrown by Scalar DB storage.

I think we don't have to check kinds of these exceptions because the exception messages are shown.
In `transfer` test, I didn't check kinds of exception except for `UnknownTransactionException`, though the test sometimes tries unneeded reconnections.

Reply:
Thank you. I removed this check.